File: backend/src/main/java/com/backend/domain/plan/service/PlanMemberService.java
package com.backend.domain.plan.service;

import com.backend.domain.member.entity.Member;
import com.backend.domain.member.service.MemberService;
import com.backend.domain.plan.dto.PlanMemberMyResponseBody;
import com.backend.domain.plan.dto.PlanMemberAddRequestBody;
import com.backend.domain.plan.dto.PlanMemberAnswerRequestBody;
import com.backend.domain.plan.dto.PlanMemberResponseBody;
import com.backend.domain.plan.entity.Plan;
import com.backend.domain.plan.entity.PlanMember;
import com.backend.domain.plan.repository.PlanMemberRepository;
import com.backend.global.exception.BusinessException;
import com.backend.global.response.ErrorCode;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;

@Slf4j
@Service
@RequiredArgsConstructor
public class PlanMemberService {
    private final PlanMemberRepository planMemberRepository;
    private final PlanService planService;
    private final MemberService memberService;

    public PlanMemberResponseBody invitePlanMember(PlanMemberAddRequestBody requestBody, long memberId) {
        PlanMember planMember = isValidInvite(requestBody, memberId);
        planMemberRepository.save(planMember);
        return new PlanMemberResponseBody(planMember);
    }

    public List<PlanMemberMyResponseBody> myInvitedPlanList(long memberPkId) {
        Member member = Member.builder().id(memberPkId).build();

        List<PlanMember> planMemberList = planMemberRepository.getPlanMembersByMember(member);
        List<PlanMemberMyResponseBody> myPlanMemberList =
                planMemberList
                        .stream()
                        .map(pm -> new PlanMemberMyResponseBody(pm))
                        .toList();
        return myPlanMemberList;
    }


    public PlanMemberResponseBody DeletePlanMember(PlanMemberAddRequestBody requestBody, long memberPkId) {
        PlanMember planMember = isValidInvite(requestBody, memberPkId);

        planMemberRepository.delete(planMember);
        return new PlanMemberResponseBody(planMember);
    }

    public PlanMemberResponseBody acceptInvitePlanMember(PlanMemberAnswerRequestBody requestBody, long memberPkId) {
        PlanMember planMember = isMyInvite(requestBody, memberPkId);
        planMember.inviteAccept();
        planMemberRepository.save(planMember);
        return new PlanMemberResponseBody(planMember);
    }

    public PlanMemberResponseBody denyInvitePlanMember(PlanMemberAnswerRequestBody requestBody, long memberPkId) {
        PlanMember planMember = isMyInvite(requestBody, memberPkId);
        planMember.inviteDeny();
        planMemberRepository.save(planMember);
        return new PlanMemberResponseBody(planMember);
    }

    private PlanMember isValidInvite(PlanMemberAddRequestBody requestBody, long memberId) {
        Plan plan = planService.getPlanById(requestBody.planId());
        if (plan.getMember().getId() != memberId) {
            throw new BusinessException(ErrorCode.NOT_MY_PLAN);
        }

        Member invitedMember = memberService.findById(requestBody.memberId());

        // 데이터 베이스 오류 처리를 서비스 로직 처리로 변경
        if(planMemberRepository.existsByMemberInPlanId(invitedMember.getId(),plan.getId())) {
            throw new BusinessException(ErrorCode.DUPLICATE_MEMBER_INVITE);
        };

        PlanMember planMember = new PlanMember(null,invitedMember,plan,null,null,0);
        return planMember;
    }

    private PlanMember isMyInvite(PlanMemberAnswerRequestBody requestBody, long memberPkId) {
        Member member = Member.builder().id(memberPkId).build();

        Plan plan = planService.getPlanById(requestBody.planId());

        if (requestBody.memberId() != member.getId()) {
            throw new BusinessException(ErrorCode.NOT_MY_PLAN);
        }

        PlanMember planMember = planMemberRepository.findById(requestBody.planMemberId()).orElseThrow(
                () -> new BusinessException(ErrorCode.NOT_FOUND_INVITE)
        );

        return planMember;
    }

    public boolean isAvailablePlanMember(long planId, Member member) {
        return planMemberRepository.existsByMemberInPlanId(member.getId(),planId);
    }
}

File: backend/src/main/java/com/backend/domain/plan/service/PlanMemberService.kt
package com.backend.domain.plan.service

import com.backend.domain.member.entity.Member
import com.backend.domain.member.entity.QMember.member
import com.backend.domain.member.service.MemberService
import com.backend.domain.plan.dto.PlanMemberAddRequestBody
import com.backend.domain.plan.dto.PlanMemberAnswerRequestBody
import com.backend.domain.plan.dto.PlanMemberMyResponseBody
import com.backend.domain.plan.dto.PlanMemberResponseBody
import com.backend.domain.plan.entity.PlanMember
import com.backend.domain.plan.entity.QPlan.plan
import com.backend.domain.plan.repository.PlanMemberRepository
import com.backend.global.exception.BusinessException
import com.backend.global.response.ErrorCode
import lombok.RequiredArgsConstructor
import lombok.extern.slf4j.Slf4j
import org.springframework.data.jpa.domain.AbstractPersistable_.id
import org.springframework.stereotype.Service
import java.util.function.Supplier

@Slf4j
@Service
@RequiredArgsConstructor
class PlanMemberService(
    private val planMemberRepository: PlanMemberRepository,
    private val planService: PlanService,
    private val memberService: MemberService
) {


    fun invitePlanMember(requestBody: PlanMemberAddRequestBody, memberPkId: Long): PlanMemberResponseBody {
        val planMember = isValidInvite(requestBody, memberPkId)
        planMemberRepository.save<PlanMember?>(planMember)
        return PlanMemberResponseBody(planMember)
    }


    fun myInvitedPlanList(memberPkId: Long): List<PlanMemberMyResponseBody> {
        // TODO : 이후 ID 값만 있는 멤버 객체를 생성하는 방법 찾기
        val member: Member = memberService.findById(memberPkId)
        val planMemberList = planMemberRepository.getPlanMembersByMember(member)
        val myPlanMemberList =
            planMemberList
                .filter { it.plan.member.id != memberPkId }
                .map{ PlanMemberMyResponseBody(it) }
                .toList()

        return myPlanMemberList
    }

    fun DeletePlanMember(requestBody: PlanMemberAddRequestBody, memberPkId: Long): PlanMemberResponseBody {
        val planMember : PlanMember = planMemberRepository
            .getMyInviteByMemberIdAndPlanId(requestBody.planId,requestBody.memberId,memberPkId)
            ?: throw BusinessException(ErrorCode.NOT_FOUND_INVITE)

        planMemberRepository.delete(planMember)
        return PlanMemberResponseBody(planMember)
    }

    fun acceptInvitePlanMember(requestBody: PlanMemberAnswerRequestBody, memberPkId: Long): PlanMemberResponseBody {
        val planMember = isMyInvite(requestBody, memberPkId)
        planMember.inviteAccept()
        planMemberRepository.save<PlanMember?>(planMember)
        return PlanMemberResponseBody(planMember)
    }

    fun denyInvitePlanMember(requestBody: PlanMemberAnswerRequestBody, memberPkId: Long): PlanMemberResponseBody {
        val planMember = isMyInvite(requestBody, memberPkId)
        planMember.inviteDeny()
        planMemberRepository.save<PlanMember?>(planMember)
        return PlanMemberResponseBody(planMember)
    }

    // 초대가 유효한지 검사합니다.
    private fun isValidInvite(requestBody: PlanMemberAddRequestBody, memberId: Long): PlanMember {
        val plan = planService.getPlanById(requestBody.planId)
        if (plan.member.id != memberId) throw BusinessException(ErrorCode.NOT_MY_PLAN) // 현재 로그인 한 사용자의 계획이 맞는지 확인합니다.

        val invitedMember = memberService.findById(requestBody.memberId) // 초대 할 사용자의 계정이 존재하는지 확인하고 해당 객체를 가져옵니다.
        // DB에 해당 초대가 이미 존재하는지 확인합니다.
        if (planMemberRepository.existsByMemberInPlanId(
                invitedMember.id?:throw BusinessException(ErrorCode.INVALID_MEMBER),
                plan.id?:throw BusinessException(ErrorCode.NOT_FOUND_PLAN))) {
            throw BusinessException(ErrorCode.DUPLICATE_MEMBER_INVITE)
        }

        val planMember = PlanMember(null, invitedMember, plan, null, null, 0)
        return planMember
    }

    // 내 초대가 맞는지 검사합니다.
    private fun isMyInvite(requestBody: PlanMemberAnswerRequestBody, memberPkId: Long): PlanMember {
        // TODO : 이후 ID 값만 있는 멤버 객체를 생성하는 방법 찾기
        val member: Member = memberService.findById(memberPkId)

        val plan = planService.getPlanById(requestBody.planId)

        if (requestBody.memberId != member.id) {
            throw BusinessException(ErrorCode.NOT_MY_PLAN)
        }

        val planMember = planMemberRepository.findById(requestBody.planMemberId).orElseThrow<BusinessException?>(
            Supplier { BusinessException(ErrorCode.NOT_FOUND_INVITE) }
        )

        return planMember
    }

    fun isAvailablePlanMember(planId: Long, memberPkId: Long): Boolean {
        return planMemberRepository.existsByPlanIdAndMemberId( planId, memberPkId)
    }

    fun isAvailableAndAcceptedPlanMember(planId: Long, memberPkId: Long): Boolean {
        return planMemberRepository.existsByPlanIdAndMemberIdAndIsConfirmed(planId, memberPkId,1);
    }

    fun getPlanMembers(planId: Long, memberPkId: Long): List<PlanMemberResponseBody> {
        val members : List<PlanMemberResponseBody> =planMemberRepository.myQueryGetPlanMembers(planId)
        return members.filter { it.isConfirmed }
    }
}
